great_expectations/core/expectation_suite.py

```diff
@@ -144,27 +145,27 @@ def add_expectation(self, expectation: _TExpectation) -> _TExpectation:
                 "and set `Expectation.id = None`."
             )
         should_save_expectation = self._has_been_saved()
-        expectation_is_unique = all(
-            expectation.configuration != existing_expectation.configuration
-            for existing_expectation in self.expectations
-        )
-        if expectation_is_unique:
+        if self._determine_if_expectation_is_unique(expectation):
             # suite is a set-like collection, so don't add if it not unique
-            self.expectations.append(expectation)
             if should_save_expectation:
-                try:
-                    expectation = self._store.add_expectation(suite=self, expectation=expectation)
-                    self.expectations[-1].id = expectation.id
-                except Exception as exc:
-                    self.expectations.pop()
-                    raise exc  # noqa: TRY201
+                expectation = self._store.add_expectation(suite=self, expectation=expectation)
+            self.expectations.append(expectation)
 
         expectation.register_save_callback(save_callback=self._save_expectation)
 
         self._submit_expectation_created_event(expectation=expectation)
 
         return expectation
 
+    def _determine_if_expectation_is_unique(self, expectation: Expectation) -> bool:
+        # Expectation is deemed unique if it is not already in the suite
+        # We do not consider the id of the expectation in this check
+        expectation_copies = copy.deepcopy(self.expectations)
```

Comment:
I'm guessing we also want to account for `rendered_content`, right?

Also does this get expensive when we have rendered content? If we want to save ourselves the deep copy (and mutation), you could do something like `model1.dict(exclude={"id"}) == model2.dict(exclude={"id"})` in a loop, or that's what chatgpt tells me at least 😄